file/src/test/java/docs/javadsl/ArchiveTest.java

```diff
@@ -188,22 +200,61 @@ public void tarReader() throws Exception {
                 1,
                 pair -> {
                   TarArchiveMetadata metadata = pair.first();
-                  Source<ByteString, NotUsed> source = pair.second();
                   Path targetFile = target.resolve(metadata.filePath());
-                  // create the target directory
-                  return Source.single(targetFile.getParent())
-                      .via(Directory.mkdirs())
-                      .runWith(Sink.ignore(), mat)
-                      .thenCompose(
-                          done ->
-                              // stream the file contents to a local file
-                              source.runWith(FileIO.toPath(targetFile), mat));
+                  if (metadata.isDirectory()) {
+                    return Source.single(targetFile)
+                        .via(Directory.mkdirs())
+                        .runWith(Sink.ignore(), mat);
+                  } else {
+                    Source<ByteString, NotUsed> source = pair.second();
+                    // create the target directory
+                    return Source.single(targetFile.getParent())
+                        .via(Directory.mkdirs())
+                        .runWith(Sink.ignore(), mat)
+                        .thenCompose(
+                            done ->
+                                // stream the file contents to a local file
+                                source
+                                    .runWith(FileIO.toPath(targetFile), mat)
+                                    .thenApply(io -> Done.done()));
+                  }
                 })
             .runWith(Sink.ignore(), mat);
     // #tar-reader
     assertThat(tar.toCompletableFuture().get(2, TimeUnit.SECONDS), is(Done.done()));
     File file = target.resolve("dir/file1.txt").toFile();
-    assertTrue(file.exists());
+    assertThat(file.exists(), is(true));
+  }
+
+  @Test
+  public void tarReaderWithEofBlocks() throws Exception {
+    ByteString tenDigits = ByteString.fromString("1234567890");
+    TarArchiveMetadata metadata1 = TarArchiveMetadata.create("file5.txt", tenDigits.length());
+    CompletionStage<ByteString> oneFileArchive =
+        Source.single(Pair.create(metadata1, Source.single(tenDigits)))
+            .via(Archive.tar())
+            // tar standard suggests two empty 512 byte blocks as EOF marker
+            .concat(Source.single(ByteString.fromArray(new byte[1024])))
```

Comment:
I checked what tar does on my machine and there the padding seems to be 20 x 512 which is somewhat suggested by this spec: https://www.gnu.org/software/tar/manual/html_node/Blocking-Factor.html.

Anyway, it shouldn't matter since any amount of data is skipped at the end.

Would be good to add an actual file created by (some version of) tar itself, or am I missing that?

Reply:
A user shared a file which blew up in Alpakka with me (it padded with 6x512 bytes), but I can't add that file to the repository.